### DISCUSSION TITLE
Compile time and Runtime/JIT Time Side Effect Factoring & Isolation, Tracking and Inference

### DISCUSSION THREAD

**Discussion by sirinath** [dotnet/runtime#74870] [Ideas]
Can you can add side effect 
- factoring & isolation 
- track and 
- inference

This will have to deal with:
- IO
- Memory operations
- Exception
- Reflection

E.g. 1. You are in a loop which does a side effect but hits out of bounds. We know before the exception there are these side affects plus the exception which is a side effect. Eliminate bound checking upto the point the exception is thrown by breaking the loop into 2 ranges. If any side of the effects are not observable from outside either by memory mutation, exception or reflection, just eliminate the whole part of code. 

E.g. 2. If a loop prints 1 to 100, instead of multiple IO operations, in absence of any other possibilities. Fuse all the IO operations into 1 operation. 

E.g. 3. If we initialise multiple array elements certain value, just initialise memory than doing this.

Simply put it, if something to not externally observable, no need to do it. If what is observed can be inferred statically or during JIT do it, Boundary can be Assembly or Application depending on context.
